nameserver/dns.go

```diff
@@ -1,117 +1,239 @@
+		ip := addr.IP4()
+		response.Answer[i] = &dns.A{Hdr: *header, A: ip}
+	}
+	shuffleAnswers(&response.Answer)
+
+	Log.Debugf("[gossipdns] response: %+v", response)
+	if err := w.WriteMsg(&response); err != nil {
+		Log.Infof("[gossipdns] error responding: %v", err)
+	}
 }
 
-// shuffleAnswers reorders answers for very basic load balancing
-func shuffleAnswers(answers []dns.RR) []dns.RR {
-	if len(answers) > 1 {
-		for i := range answers {
-			j := rand.Intn(i + 1)
-			answers[i], answers[j] = answers[j], answers[i]
+func (d *DNSServer) handleReverse(client *dns.Client) func(dns.ResponseWriter, *dns.Msg) {
+	return func(w dns.ResponseWriter, req *dns.Msg) {
+		Log.Debugf("[gossipdns] request: %+v", *req)
+		if len(req.Question) != 1 || req.Question[0].Qtype != dns.TypePTR {
+			badDNSRequest(req, w)
+			return
+		}
+
+		ipStr := strings.TrimSuffix(req.Question[0].Name, "."+reverseDNSdomain)
+		ip, err := address.ParseIP(ipStr)
+		if err != nil {
+			badDNSRequest(req, w)
+			return
+		}
+
+		hostname, err := d.ns.ReverseLookup(ip.Reverse())
+		if err != nil {
+			d.handleRecursive(client)(w, req)
+			return
+		}
+
+		response := dns.Msg{}
+		response.RecursionAvailable = true
+		response.Authoritative = true
+		response.SetReply(req)
+
+		header := &dns.RR_Header{
+			Name:   req.Question[0].Name,
+			Rrtype: dns.TypePTR,
+			Class:  dns.ClassINET,
+			Ttl:    d.ttl,
+		}
+
+		response.Answer = []dns.RR{&dns.PTR{
+			Hdr: *header,
+			Ptr: hostname,
+		}}
+
+		Log.Debugf("[gossipdns] response: %+v", response)
+		if err := w.WriteMsg(&response); err != nil {
+			Log.Infof("[gossipdns] error responding: %v", err)
 		}
 	}
+}
 
-	return answers
+func (d *DNSServer) handleRecursive(client *dns.Client) func(dns.ResponseWriter, *dns.Msg) {
+	return func(w dns.ResponseWriter, r *dns.Msg) {
+		Log.Debugf("[gossipdns] request: %+v", *r)
+		for _, server := range d.upstream.Servers {
+			reply, _, err := client.Exchange(r, fmt.Sprintf("%s:%s", server, d.upstream.Port))
+			if err != nil {
+				Log.Debugf("[gossipdns] network error trying %s (%s)", r.MsgHdr.Id, server, err)
+				continue
+			}
+			if reply != nil && reply.Rcode != dns.RcodeSuccess {
+				Log.Debugf("[gossipdns] failure reported by %s", r.MsgHdr.Id, server)
+				continue
+			}
+			Log.Debugf("[gossipdns] response: %+v", *reply)
+			if err := w.WriteMsg(reply); err != nil {
+				Log.Infof("[gossipdns] error responding: %v", err)
+			}
+			return
+		}
+
+		badDNSRequest(r, w)
+	}
 }
 
-// only take the first `num` answers
-func pruneAnswers(answers []dns.RR, num int) []dns.RR {
-	if num > 0 && len(answers) > num {
-		// TODO: we should have some prefer locally-introduced answers, etc...
-		return answers[:num]
+func shuffleAnswers(answers *[]dns.RR) {
+	if len(*answers) <= 1 {
+		return
+	}
+
+	for i := range *answers {
+		j := rand.Intn(i + 1)
+		(*answers)[i], (*answers)[j] = (*answers)[j], (*answers)[i]
 	}
-	return answers
+}
+
+func (d *DNSServer) HandleHTTP(router *mux.Router) {
```

Comment:
Yes, but at the cost of having a `HandleHTTP` inside the DNS server and dealing with HTTP there... It would be better to just pass the `domain` when you create the HTTP server.